File: api/src/main/java/dev/akif/crud/simpler/SimplerMapper.java
package dev.akif.crud.simpler;

import dev.akif.crud.CRUDDTOMapper;
import dev.akif.crud.CRUDMapper;

import java.io.Serializable;

/**
 * Simpler variant of {@link CRUDMapper} where create/update models and DTOs are just models,
 * which is also a {@link CRUDDTOMapper}
 *
 * @param <I> ID type of the entity
 * @param <E> Entity type
 * @param <M> Model type
 */
public interface SimplerMapper<
        I extends Serializable,
        E extends SimplerEntity<I, E>,
        M extends SimplerModel<I, E, M>> extends CRUDMapper<I, E, M, M, M>, CRUDDTOMapper<I, E, M, M, M, M, M, M> {
    @Override
    default M modelToDTO(final M model) {
        return model;
    }
}

File: api/src/main/kotlin/dev/akif/crud/simpler/SimplerMapper.kt
package dev.akif.crud.simpler

import dev.akif.crud.CRUDDTOMapper
import dev.akif.crud.CRUDMapper
import dev.akif.crud.common.Parameters
import java.io.Serializable

/**
 * Simpler variant of [CRUDMapper] where create/update models and DTOs are just models,
 * which is also a [CRUDDTOMapper]
 *
 * @param I Id type of the entity
 * @param E Entity type of the data which is a [SimplerEntity]
 * @param M Model type of the data which is a [SimplerModel]
 */
interface SimplerMapper<
        I : Serializable,
        E : SimplerEntity<I>,
        M : SimplerModel<I>> : CRUDMapper<I, E, M, M, M>, CRUDDTOMapper<I, M, M, M, M, M, M> {
    override fun modelToDTO(model: M, parameters: Parameters): M = model

    override fun createDTOToCreateModel(createDTO: M, parameters: Parameters): M = createDTO

    override fun updateDTOToUpdateModel(updateDTO: M, parameters: Parameters): M = updateDTO
}
